### PR TITLE
Image upload via POST

### PR DESCRIPTION
**Pull request by qqTYXn7**

### CODE REVIEW

File: src/api/controllers/ItemImageController.ts [hunk 1]
```diff
@@ -29,14 +37,88 @@ export class ItemImageController {
...
+                                originalName: imageFile.originalname
+                            }
+                        } as ItemImageUpdateRequest);
+                }
+            }
+        }
+        throw new MessageException('Was expecting an image file called "image".');
+    }
+
+    public makeid(): string {
```
Comment (kewde) [inline src/api/controllers/ItemImageController.ts]: What is this ID generation used for? Is this a filename or a field in the DB?

File: src/api/controllers/ItemImageController.ts [hunk 2]
```diff
@@ -29,14 +37,88 @@ export class ItemImageController {
...
+            if ( file.fieldname === 'image' ) {
+                const imageFile = file;
+
+                // Read the file data in
+                const dataStr = fs.readFileSync(imageFile.path, 'base64');
+                // this.log.error('dataStr = ' + dataStr);
+
+                // find listing item template
+                const listingItemTemplate = await this.listingItemTemplateService.findOne(templateId);
+                this.log.debug('imageFile.mimetype = ' + imageFile.mimetype);
```
Comment (xludx) [inline src/api/controllers/ItemImageController.ts]: listingItemTemplate will be the same for all images, but you are fetching it over and over again in a loop.

File: src/api/controllers/ItemImageController.ts [hunk 3]
```diff
@@ -29,14 +37,88 @@ export class ItemImageController {
...
+                            }
+                        } as ItemImageUpdateRequest);
+                } else {
+                    const itemImageId = itemImage.id;
+                    // this.log.debug('itemImage = ' + JSON.stringify(itemImage));
+                    this.log.debug('itemImageId = ' + itemImageId);
+
+                    // create item images
+                    return await this.itemImageService.update(
+                        itemImageId,
```
Comment (xludx) [inline src/api/controllers/ItemImageController.ts]: if item allready has one image then it's updated, why?

File: src/api/controllers/ItemImageController.ts (hunk 2 above)
Comment (qqTYXn7) [reply src/api/controllers/ItemImageController.ts]: This is true, but we only expect one file to be uploaded with the fieldname 'image', so shouldn't happen multiple times

File: src/api/controllers/ItemImageController.ts (hunk 3 above)
Comment (qqTYXn7) [reply src/api/controllers/ItemImageController.ts]: I don't understand.
Do you want it so if an image already exists we just disregard the upload and use what we already have?

File: src/api/controllers/ItemImageController.ts (hunk 1 above)
Comment (qqTYXn7) [reply src/api/controllers/ItemImageController.ts]: That was just a placeholder.
Now contains the name of the uploaded file

File: src/api/controllers/ItemImageController.ts [hunk 4]
```diff
@@ -18,46 +25,93 @@ export class ItemImageController {
...
+        if (!templateIdStr) {
+            throw new MessageException('ListingItemTemplate id can not be null.');
+        } else {
+            templateId = parseInt(templateIdStr, 10);
+            if ( !templateId ) {
+                throw new MessageException('ListingItemTemplate id must be an integer.');
+            }
+        }
+
+        // this.log.debug('templateId: ' + templateId);
```
Comment (xludx) [inline src/api/controllers/ItemImageController.ts]: everything below this looks like code that should be in service

File: src/api/controllers/ItemImageController.ts [hunk 5]
```diff
@@ -18,46 +25,93 @@ export class ItemImageController {
...
+        // also check listingItemTemplate id present in params
+        let templateId: number;
+        if (!templateIdStr) {
+            throw new MessageException('ListingItemTemplate id can not be null.');
+        } else {
+            templateId = parseInt(templateIdStr, 10);
+            if ( !templateId ) {
+                throw new MessageException('ListingItemTemplate id must be an integer.');
+            }
+        }
```
Comment (xludx) [inline src/api/controllers/ItemImageController.ts]: pass the execution to service and do requestbody validation for values using annotations @IsNumber, @IsInt, @IsNotEmpty, etc and get rid of these unnecessary lines of code.

File: src/api/factories/ImageFactory.ts [hunk 6]
```diff
@@ -57,6 +57,8 @@ export class ImageFactory {
             protocol: originalImageData.protocol,
```
Comment (xludx) [inline src/api/factories/ImageFactory.ts]: all the other factories use getModel method to get/create a Model (or actually CreateRequest) out of the data and getMessage to get the Message to send to the network. even though this isn't exactly the definition of Factory pattern and requires some refactoring in the future, and I should have propably done an interface for this, we should use similar naming everywhere so we don't need to guess what the method does. In this case I dont really need to guess, but still we need to be consistent or in a few months we have a codebase where you need to read every single line of code in a function to understand wtf its about or for.

File: src/api/services/ItemImageHttpUploadService.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    public async httpPostImageUpload(@request(ImagePostUploadRequest) uploadRequest: ImagePostUploadRequest): Promise<any> {
+        const listItems: any[] = [];
+        const listingItemTemplate: ListingItemTemplate = await this.listingItemTemplateService.findOne(uploadRequest.id);
+        for ( const file of uploadRequest.request.files ) {
+            const tmpImage = await this.itemImageService.createFile(uploadRequest.id, file, listingItemTemplate);
+            const imageDatas = tmpImage.ItemImageDatas;
+            for ( const i in imageDatas ) {
+                if ( i ) {
+                    const tmpImageData: any = imageDatas[i];
+                    tmpImageData.data = 'http://../../../item-image-data/' + tmpImageData.id;
```
Comment (xludx) [inline src/api/services/ItemImageHttpUploadService.ts]: add correct url, process.env.APP_HOST + ...

Comment (xludx) [reply src/api/services/ItemImageHttpUploadService.ts]: also the link should be in the dataid field and protocol should be set as http I think

      {
        "hash": "f84f0..702",
        "data": [
          {
            "protocol": "IPFS",
            "id": "QmUwHMFY9GSiKgjqyZpgAv2LhBrh7GV8rtLuagbry9wmMU"
          },
          {
            "protocol": "HTTPS",
            "id": "https://link.to/image1"
          },
          {
            "protocol": "ONION",
            "id": "https://3g2upl4pq6kufc4m.onion/assets/logo_homepage.normal.v107.png"
          },

File: src/api/services/ItemImageHttpUploadService.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        const listingItemTemplate: ListingItemTemplate = await this.listingItemTemplateService.findOne(uploadRequest.id);
+        for ( const file of uploadRequest.request.files ) {
+            const tmpImage = await this.itemImageService.createFile(uploadRequest.id, file, listingItemTemplate);
+            const imageDatas = tmpImage.ItemImageDatas;
+            for ( const i in imageDatas ) {
+                if ( i ) {
+                    const tmpImageData: any = imageDatas[i];
+                    tmpImageData.data = process.env.APP_HOST + (process.env.APP_PORT ? ':' + process.env.APP_PORT: '') + '/api/item-image-data/' + tmpImageData.id;
+                    tmpImageData.dataId = tmpImageData.data;
+                    tmpImageData.protocol = 'HTTP';
```
Comment (kewde) [inline src/api/services/ItemImageHttpUploadService.ts]: Should be local no?

**Pull request merged**